### PR TITLE
importccl: support AUTO_INCREMENT in mysqldump

### LINKED ISSUE #28711: importccl: mysqldump import ignores AUTO_INCREMENT
in the dumped data, `mysqldump` does give us a hint as to where the sequence should start, so we just need to do the same thing we do for pgdump:
```
CREATE TABLE `test_table` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `jsondata` json DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=5 DEFAULT CHARSET=utf8;
```

### PR DESCRIPTION
**Pull request by dt**
This adds support for automatically creating a sequence and using it in the default value of AUTO_INCREMENT mysql columns.
This functions similarly to the pgdump functionality with a couple differences:
  - the mysqldump does not explicitly create a sequence or reference it, so we must make one and use that
  - the mysqldump format includes the next value in the table definition, so we need to capture it when we read the table,
    and then use it to set the sequence later.

Fixes #28711.

Release note: none.

### CODE REVIEW

**Review by knz — COMMENTED** [review]
I am surprised by the lack of release notes.

:lgtm: 

> Reviewed 4 of 4 files at r1, 2 of 2 files at r2, 12 of 13 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28761#-:-LK7Kg-YBDMFnroD-yEf:b-us4ab)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by dt** [comment]
@knz all of the MYSQLDUMP support is new, so no changes to user-facing released functionality.

File: pkg/ccl/importccl/read_import_mysql.go [hunk 1]
```diff
@@ -288,15 +289,41 @@ func mysqlTableToCockroach(
...
+			i, err := strconv.Atoi(strings.TrimPrefix(opt, seqOpt))
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "parsing AUTO_INCREMENT value")
+			}
+			i64 := int64(i)
+			seqVals[id] = i64
+			priv := sqlbase.NewDefaultPrivilegeDescriptor()
+			opts := tree.SequenceOptions{{Name: tree.SeqOptStart, IntVal: &i64}}
+			seqName = name + "_auto_inc"
+			desc, err := sql.MakeSequenceTableDesc(seqName, opts, parentID, id, time, priv, nil)
```
Comment (mjibson) [inline pkg/ccl/importccl/read_import_mysql.go]: add comment for the nil

File: pkg/ccl/importccl/read_import_mysql.go [hunk 2]
```diff
@@ -306,6 +333,16 @@ func mysqlTableToCockroach(
 		if err != nil {
 			return nil, nil, err
 		}
+		if raw.Type.Autoincrement {
+			if seqName == "" {
+				return nil, nil, errors.Errorf("column %q specifies AUTO_INCREMENT but table options did not include it", def.Name)
+			}
+			expr, err := parser.ParseExpr(fmt.Sprintf("nextval('%s':::STRING)", seqName))
```
Comment (mjibson) [inline pkg/ccl/importccl/read_import_mysql.go]: is it possible for seqName to contain a single quote? Seems possible if the table name contains a single quote?

File: pkg/ccl/importccl/read_import_mysql.go [hunk 3]
```diff
@@ -294,55 +294,42 @@ func mysqlTableToCockroach(
 	}
 
 	time := hlc.Timestamp{WallTime: evalCtx.GetStmtTimestamp().UnixNano()}
-	priv := sqlbase.NewDefaultPrivilegeDescriptor()
-	desc := sql.InitTableDescriptor(id, parentID, name, time, priv)
 
-	colNames := make(map[string]string)
+	stmt := &tree.CreateTable{
+		Table: tree.NormalizableTableName{tree.NewTableName("", tree.Name(name))},
```
[the remark this answers is not in the document]
Comment (knz) [reply pkg/ccl/importccl/read_import_mysql.go]: yes correct

File: pkg/ccl/importccl/read_import_mysql.go (hunk 2 above)
Comment (dt) [reply pkg/ccl/importccl/read_import_mysql.go]: Added another commit that uses [`CompliantName`](https://github.com/vitessio/vitess/blob/master/go/vt/sqlparser/ast.go#L3597) everywhere, which replaces any non-letter, and any non-digit or leading digit with `_` as far as I can tell.

File: pkg/ccl/importccl/read_import_mysql.go (hunk 1 above)
Comment (dt) [reply pkg/ccl/importccl/read_import_mysql.go]: done.

File: pkg/ccl/importccl/read_import_mysql.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dt) [reply pkg/ccl/importccl/read_import_mysql.go]: Done.

**Comment by dt** [comment]
This is ready for another look -- hoping to merge it by tomorrow.

**Comment by dt** [comment]
bors r+

merging on knz's lgtm to make branch cutting day, but if there are any bugs I missed happy to fix in follow up

**Pull request merged**